### DISCUSSION TITLE
Unhandled Runtime Error Error: Functions cannot be passed directly to Client Components unless you explicitly expose it by marking it with "use server".

### DISCUSSION THREAD

**Discussion by aamaltf1** [vercel/next.js#47846] [Help]
### Summary

Provide environment information
Operating System:
Platform: windows
Arch: x64
Node: 16.17.1
npm: 9.5.0
Yarn: N/A
Relevant packages:
next: 13.2.4
eslint-config-next: 13.2.4
react: 18.2.0
react-dom: 18.2.0

What browser are you using? (if relevant)
Chrome

How are you deploying your application? (if relevant)
No response

Describe the Bug

passing a function as a prop from a server side component to client side component
```

export default function Page() {
  let term;

  const setTerm= (term: string = 'health'): void=>{
     "use server"
     term = term
  }
  const component = getTenders(term)
    .then((res) => res)
    .then((data) => {

      const cardData = getCardData(data.results);
      const metadata = { ...data };
      delete metadata.results;
      return <Container setTerm={setTerm} metadata={metadata} tenders={[...cardData]} />;
    });
  return component;
}

```
Page is a server side component and container is a client side component
To Reproduce
Click the link in the browser to navigate to /notes/1

**Comment by icyJoseph** [comment]
You can only pass serializable props: https://beta.nextjs.org/docs/rendering/server-and-client-components#passing-props-from-server-to-client-components-serialization

> Props passed from the Server Components to Client components need to be [serializable](https://developer.mozilla.org/en-US/docs/Glossary/Serialization). This means that values such as functions, Dates, etc, cannot be passed directly to client components.

Roughly, and simplifying a bit here, it means that it must be able to be destroyed, when going through some boundary, the HTTP protocol, and be reconstructed on the other side. Functions can't do that.

**Reply by Gyurmatag** [reply]
Then how should we do this if we have a click event inside a client side component and we want to react to that in a server side mother component to mutate an API endpoint?

**Reply by okormorupraise** [reply]
pls have you figured it out?

**Reply by icyJoseph** [reply]
The "issue" is that the Next.js interpretation/implementation of the React Server Components spec doesn't allow for this, at this current time anyway. 

I've solved this with queries, see here: https://github.com/vercel/next.js/discussions/47227#discussioncomment-5342500

However, as far as I understand this is but an implementation detail that's missing. Do remember that this is all beta technology, and at a very early stage.

There's an experimental framework called WakuWork: https://github.com/dai-shi/wakuwork which allows for this pattern.

Think of the implications of what you are trying to do. You'd want to pass a function from the server to the client:

- JavaScript context is lost, as functions cannot be rehydrated (strictly speaking)
- You'd need to fly a request from the client to the server with the updated data, who controls this async work?

Well WakuWork has a solution from that, as per RSC spec if I am reading things correctly. Hopefully that comes to Next.js sooner rather than later.

In WakuWork you define a server side function and pass it as props to the client component, the framework extract this function and instead creates an action to be run on the server at runtime, instead an action trigger is given to the client. When the trigger is executed client side a request is made to the server and the result is used to update the DOM.

Here's some material to explore further: https://youtu.be/ciQdQkT0bZk

**Reply by codinginflow** [reply]
> You can only pass serializable props: https://beta.nextjs.org/docs/rendering/server-and-client-components#passing-props-from-server-to-client-components-serialization
> 
> > Props passed from the Server Components to Client components need to be [serializable](https://developer.mozilla.org/en-US/docs/Glossary/Serialization). This means that values such as functions, Dates, etc, cannot be passed directly to client components.
> 
> Roughly, and simplifying a bit here, it means that it must be able to be destroyed, when going through some boundary, the HTTP protocol, and be reconstructed on the other side. Functions can't do that.

Then how are we supposed to pass server actions to client components as described here?
https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions#props

**Comment by okormorupraise** [comment]
pls have you figured it out?

**Comment by mod7ex** [comment]
maybe this will be solved in comming versions, i think it's not easy to serialize a function)) well this limits us on how we can use React Power

**Reply by icyJoseph** [reply]
The wakuwork experimental framework has accomplished this with triggers/actions: https://youtu.be/ciQdQkT0bZk this YT videp explains it well. 

Hopefully some version of that can be implemented here as well.

Though there's quite a bit to consider when this client side actions trigger server work, while keeping the UI Interactive. It's not trivial!

Link to wakuwork: https://github.com/dai-shi/wakuwork

**Comment by meshackm** [comment]
I was running into the same error when using react query. I was putting the QueryClientProvider in the layout.tsx in app folder. I tried another way I learned from @developedbyed in one of his tutorials. I created a separate QueryWrapper.tsx file and made it a client component using "use client"; then imported it into layout.tsx 

QueryWrapper below:

```
"use client";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

interface Props {
  children: React.ReactNode;
}

const queryClient = new QueryClient();

const QueryWrapper = ({ children }: Props) => (
  <QueryClientProvider client={queryClient}>
    {children}
  </QueryClientProvider>
);
export default QueryWrapper;
```

After importing it into layout.tsx wrap children like this:

```
<html lang="en" data-theme="light">
      <head></head>
      <body>
        <QueryWrapper>
          {children}
          <ReactQueryDevtools />
        </QueryWrapper>
      </body>
    </html>
```

**Comment by fzsf163** [comment]
_**Unhandled Runtime Error
Error: Functions cannot be passed directly to Client Components unless you explicitly expose it by marking it with "use server".
  <... src=... width={350} height={350} alt=... loader={function}>**_

_note: this error did not occur when I used same methods @/components/Products.tsx, and imported a component named "ProductCard" inside it, it only happened when tried to do it in a dynamic page with imported component "singleProduct"> see codes below for details_

### TLDR:
**there was a component that would display a single product. That function was giving this error.** 
**Solution: "use client" where the error was occuring**



This **"SingleProduct"** component (which takes a prop) was called inside a page name **"ProductDetailView**" [a dynamic route] which by default should be server rendered , where the single product was fetched from api call. 

After doing every thing else like enabling server action, I just wrote "use client" on the single product component and all error went way. 

I believe mine happened because I was fetching the data in a dynamic page which was server side but the component was client side. But not sure. if any one know whats going on please let me know. [btw I am noob]

```
"dependencies": {
    "@radix-ui/react-dialog": "^1.0.4",
    "@radix-ui/react-navigation-menu": "^1.1.3",
    "@radix-ui/react-slot": "^1.0.2",
    "@types/node": "20.4.5",
    "@types/react": "18.2.17",
    "@types/react-dom": "18.2.7",
    "autoprefixer": "10.4.14",
    "axios": "^1.4.0",
    "class-variance-authority": "^0.7.0",
    "clsx": "^2.0.0",
    "eslint": "8.45.0",
    "eslint-config-next": "13.4.12",
    "lucide-react": "^0.263.1",
    "next": "13.4.12",
    "postcss": "8.4.27",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "tailwind-merge": "^1.14.0",
    "tailwindcss": "3.3.3",
    "tailwindcss-animate": "^1.0.6",
    "typescript": "5.1.6"
  }
```



All codes are as it is.

for ui > shadcn/ui
_route > @/ProductDetailView/[id]/page.tsx_

```
// product component
// by deafult should be server

import { ProductProps } from "@/components/Products";
import axios from "axios";
import SingleProduct from "@/components/SingleProduct";


export async function getSingleProducts ( id: string ) {

    const res = await axios.get( `https://fakestoreapi.com/products/${ id }`, {
        // timeout after 5 seconds
        // signal: AbortSignal.timeout( 5000 ),
    } )
        .then( response => { return response.data } )
        .catch( err => alert( err ) )

    return res;
}

export default async function ProductDetailView ( { params }: { params: { id: string } } ) {

    const product: ProductProps = await getSingleProducts( params.id );

    return (
        <main className="text-center text-gray-200">
            <SingleProduct props={ product }></SingleProduct>
        </main>
    )
}

```

_route > @/components/SignleProduct_

```
// adding use client solves the issue.
"use client"
import { ProductProps } from "@/components/Products";
import {
    Card,
    CardContent,
    CardDescription,
    CardFooter,
    CardHeader,
    CardTitle,
} from "@/components/ui/card"
import Image from "next/image";

export default function SingleProduct ( { props }: { props: ProductProps } ) {


    const imgLink = ( src: string ) => {
        const image = src.substring( src.indexOf( "img" ) );
        return image
    }

    // beautifully parse it in a variable
    const imgSrc = imgLink( props.image );

    // now the loader
    // always destructure src or an object will be returned 
    const imageLoader = ( { src }: any ) => {
        const imgString = `https://fakestoreapi.com/${ src }`
        return imgString;
    }


    return (
        <main className="text-center text-gray-200">
            <Card>
                <CardHeader>
                    <CardTitle>{ props.title }</CardTitle>
                    <CardDescription>{ props.description }</CardDescription>
                </CardHeader>
                <CardContent>
                    <Image
                        src={ imgSrc }
                        width={ 350 }
                        height={ 350 }
                        alt={ props.title }
                        loader={ imageLoader }
                    >

                    </Image>
                </CardContent>
                <CardFooter>
                    <p>{ props.price }</p>
                    <p>{ props.category }</p>
                </CardFooter>
            </Card>

        </main>
    )
}
```

**The error did not happen when I did this.**

_route > @/component/Products_

```

// wanted to show all products here
import axios from "axios"
import ProductCard from "./ProductCard"

export type ProductProps = {
    id: number,
    title: string,
    price: number,
    category: string,
    description: string,
    image: string
}

export async function getProducts () {


    const res = await axios.get( "https://fakestoreapi.com/products", {
        // timeout after 5 seconds
        // signal: AbortSignal.timeout( 5000 ),
    } )
        .then( response => { return response.data } )
        .catch( err => alert( err ) )

    return res;
}

export default async function Products () {


    const products: ProductProps[] = await getProducts();

    return (
        <div>
            <h1 className="text-center text-3xl font-extrabold text-gray-50">
                Products Count : { products.length };
            </h1>

            <div className="m-5 grid grid-cols-2 gap-4 justify-center items-center">
                {
                    products.map( product => {
                        return (
                            <div key={ product.id } className="col-auto ">
                                <ProductCard props={ product }></ProductCard>
                            </div>
                        )
                    } )
                }
            </div>
        </div>
    )

}
```

**Product Card**

```

// this is product card

import {
    Card,
    CardContent,
    CardDescription,
    CardFooter,
    CardHeader,
    CardTitle,
} from "@/components/ui/card"
import { ProductProps } from "./Products"
import Link from "next/link"
import Image from "next/image"


export default function ProductCard ( { props }: { props: ProductProps } ) {

    // an elaborated way to get an external image to display
    // if we do this then we don't need to add config inside next.config.js
    // first get the last part of the url as src

    const imgLink = ( src: string ) => {
        const image = src.substring( src.indexOf( "img" ) );
        return image
    }

    // beautifully parse it in a variable
    const imgSrc = imgLink( props.image );

    // now the loader
    // always destructure src or an object will be returned 
    const imageLoader = ( { src }: any ) => {
        const imgString = `https://fakestoreapi.com/${ src }`
        return imgString;
    }

    // clip the description

    const textClip = ( src: string ) => {
        const text = src.substring( 0, 100 );
        return text
    }


    //  parse it in a variable
    const textClipped = textClip( props.description );

    // clip the title

    const titleClip = ( src: string ) => {
        const title = src.substring( 0, 100 );
        return title
    }


    //  parse it in a variable
    const titleClipped = titleClip( props.title );

    return (

        <Card className="flex justify-center items-center h-[300px] bg-slate-950 text-slate-200">
            <CardContent className="flex justify-center items-center">
                <div className="w-auto h-auto pt-6">
                    <Image
                        loader={ imageLoader }
                        src={ `${ imgSrc }` }
                        alt={ props.title }
                        width={ 340 }
                        height={ 340 }
                        layout="responsive"
                        style={ { objectFit: "contain" } }
                    >

                    </Image>
                </div>
            </CardContent>
            <section className=" flex items-start justify-center flex-col">
                <CardHeader className="pb-1 pt-0">
                    <CardTitle className="text-start text-lg">

                        {/* <LinkIcon size={ 40 } spacing={ 10 }></LinkIcon> */ }

                        <Link href={ `/ProductDetailView/${ props.id }` } target="_blank">
                            <p>{ titleClipped }...</p>
                        </Link>

                    </CardTitle>
                    <CardDescription>{ textClipped }...</CardDescription>
                </CardHeader>
                <CardContent className="p-0 ps-6">
                    <p>Category: { props.category }</p>
                </CardContent>
                <CardFooter className="p-0 ps-6" >
                    <p>${ props.price }</p>
                </CardFooter>
            </section>
        </Card>

    )
}

```

**Comment by hp41299889** [comment]
Well,I think I came up with a more intuitive solution.

If the function can not pass from the server component to the client component because function is not serializable,then I pass a serializable(like string or json) to client,and use ifelse or switch case in the client component to determine which function I should use.

In my case:

the server component
```
const clientMetadata: TableMetadata[] = [
  { key: "name", label: "姓名" },
  { key: "phone", label: "連絡電話" },
  {
    key: "birth",
    label: "出生日期",
    preProcess: "toLocaleDate",
  },
  { key: "address", label: "地址" },
  { key: "email", label: "E-mail" },
  { key: "note", label: "備註" },
  {
    key: "createdAt",
    label: "建立時間",
    preProcess: "toLocaleDateTime",
  },
  {
    key: "updatedAt",
    label: "更新時間",
    preProcess: "toLocaleDateTime",
  },
];
```

the client component
```
  const preDisplay = (preProcess: PreProcess, key: string) => {
    switch (preProcess) {
      case "toLocaleDate": {
        return toLocaleDate(key);
      }
      case "toLocaleDateTime": {
        return toLocaleDateTime(key);
      }
    }
  };
```


I don't know what the downside is,but it works.

**Comment by vtgen** [comment]
after some frustratioin, i am posting here iwth same issue;  i have essentialized it down to this

page.tsx
```tsx
"use server";
import Profile from './Profile';


export default async function Page() {

  return (
      <Profile
        getCoach={async function foo(){
          "use server"
          return null;
        }}
      />
  );
}
```

Profile.tsx
```tsx
'use client';

export interface ProfileProps {
  getCoach: () => Promise<null>;
}
const Profile = (props: ProfileProps) => {
  return <>hello{props.getCoach()}</>;
};

export default Profile;   
```

> 
> Error: Functions cannot be passed directly to Client Components unless you explicitly expose it by marking it with "use server".
>   <... getCoach={function}>
>                 ^^^^^^^^^^
>
